crypto/async/async.c

```diff
@@ -369,32 +368,41 @@ int ASYNC_init_thread(size_t max_size, size_t init_size)
 
     return 1;
 err:
-    async_free_pool_internal(pool);
+    async_empty_pool(pool);
+    sk_ASYNC_JOB_free(pool->jobs);
+    OPENSSL_free(pool);
     return 0;
 }
 
-static void async_free_pool_internal(async_pool *pool)
+void async_delete_thread_state(void)
 {
-    if (pool == NULL)
-        return;
+    async_pool *pool = (async_pool *)CRYPTO_THREAD_get_local(&poolkey);
 
-    async_empty_pool(pool);
-    sk_ASYNC_JOB_free(pool->jobs);
-    OPENSSL_free(pool);
-    CRYPTO_THREAD_set_local(&poolkey, NULL);
+    if (pool != NULL) {
+        async_empty_pool(pool);
+        sk_ASYNC_JOB_free(pool->jobs);
+        OPENSSL_free(pool);
+        CRYPTO_THREAD_set_local(&poolkey, NULL);
+    }
     async_local_cleanup();
     async_ctx_free();
 }
 
 void ASYNC_cleanup_thread(void)
 {
-    async_free_pool_internal((async_pool *)CRYPTO_THREAD_get_local(&poolkey));
+    if (!OPENSSL_init_crypto(OPENSSL_INIT_ASYNC, NULL))
+        return;
+
+    async_delete_thread_state();
 }
 
 ASYNC_JOB *ASYNC_get_current_job(void)
 {
     async_ctx *ctx;
 
+    if (!OPENSSL_init_crypto(OPENSSL_INIT_ASYNC, NULL))
+        return 0;
```

Comment:
Should return NULL rather than 0 as it's returning a pointer.

Reply:
Yes.  copied from somewhere else...